pilot/pkg/bootstrap/certcontroller.go

```diff
@@ -133,36 +130,31 @@ func (s *Server) initDNSCerts(hostname, customHost, namespace string) error {
 		names = append(names, name)
 	}
 
-	var certChain, keyPEM []byte
+	var certChain, keyPEM, caBundle []byte
 	var err error
 	if features.PilotCertProvider.Get() == KubernetesCAProvider {
 		log.Infof("Generating K8S-signed cert for %v", names)
 		certChain, keyPEM, _, err = chiron.GenKeyCertK8sCA(s.kubeClient.CertificatesV1beta1().CertificateSigningRequests(),
 			strings.Join(names, ","), hostnamePrefix+".csr.secret", namespace, defaultCACertPath)
-
-		s.caBundlePath = defaultCACertPath
+		if err != nil {
+			return fmt.Errorf("failed genrating ker cert by k8s: %v", err)
+		}
+		caBundle, err = ioutil.ReadFile(defaultCACertPath)
+		if err != nil {
+			return fmt.Errorf("failed reading %s: %v", defaultCACertPath, err)
+		}
 	} else if features.PilotCertProvider.Get() == IstiodCAProvider {
 		log.Infof("Generating istiod-signed cert for %v", names)
 		certChain, keyPEM, err = s.CA.GenKeyCert(names, SelfSignedCACertTTL.Get(), false)
+		if err != nil {
+			return fmt.Errorf("failed generating istiod key cert %v", err)
+		}
 
 		signingKeyFile := path.Join(LocalCertDir.Get(), "ca-key.pem")
 		// check if signing key file exists the cert dir
 		if _, err := os.Stat(signingKeyFile); err != nil {
 			log.Infof("No plugged-in cert at %v; self-signed cert is used", signingKeyFile)
-
-			// When Citadel is configured to use self-signed certs, keep a local copy so other
-			// components can load it via file (e.g. webhook config controller).
-			if err := os.MkdirAll(dnsCertDir, 0o700); err != nil {
-				return err
-			}
-			// We have direct access to the self-signed
-			internalSelfSignedRootPath := path.Join(dnsCertDir, "self-signed-root.pem")
-
-			rootCert := s.CA.GetCAKeyCertBundle().GetRootCertPem()
-			if err = ioutil.WriteFile(internalSelfSignedRootPath, rootCert, 0o600); err != nil {
-				return err
-			}
-
+			caBundle = s.CA.GetCAKeyCertBundle().GetRootCertPem()
 			s.addStartFunc(func(stop <-chan struct{}) error {
 				go func() {
 					for {
```

Comment:
Do we need to poll? Couldn't we just trigger the update whenever s.CA root cert is updated?

Reply:
Currently we have to, because there is no trigger mechanism, i can add a TODO here.